// neo4j_service.proto - Neo4j Graph Database gRPC Service
// Powerful graph database for connected data and relationship queries
syntax = "proto3";

package isa.neo4j;

option go_package = "github.com/isa-cloud/isa_cloud/api/proto/neo4j";

import "common.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";

// ========================================
// Neo4j Graph Database Service
// ========================================
// Provides centralized graph database access with:
// - Cypher query language support
// - Node and relationship management
// - Graph algorithms (shortest path, PageRank, etc.)
// - Transaction support
// - Full-text and spatial search

service Neo4jService {
    // ========== Cypher Query Operations ==========

    // RunCypher - Execute Cypher query
    rpc RunCypher(RunCypherRequest) returns (RunCypherResponse);

    // RunCypherRead - Execute read-only Cypher query
    rpc RunCypherRead(RunCypherReadRequest) returns (RunCypherReadResponse);

    // RunCypherWrite - Execute write Cypher query
    rpc RunCypherWrite(RunCypherWriteRequest) returns (RunCypherWriteResponse);

    // RunCypherBatch - Execute multiple Cypher queries in batch
    rpc RunCypherBatch(RunCypherBatchRequest) returns (RunCypherBatchResponse);

    // ========== Node Operations ==========

    // CreateNode - Create a new node
    rpc CreateNode(CreateNodeRequest) returns (CreateNodeResponse);

    // GetNode - Retrieve node by ID
    rpc GetNode(GetNodeRequest) returns (GetNodeResponse);

    // UpdateNode - Update node properties
    rpc UpdateNode(UpdateNodeRequest) returns (UpdateNodeResponse);

    // DeleteNode - Delete node
    rpc DeleteNode(DeleteNodeRequest) returns (DeleteNodeResponse);

    // MergeNode - Create or update node (MERGE)
    rpc MergeNode(MergeNodeRequest) returns (MergeNodeResponse);

    // FindNodes - Find nodes by label and properties
    rpc FindNodes(FindNodesRequest) returns (FindNodesResponse);

    // ========== Relationship Operations ==========

    // CreateRelationship - Create relationship between nodes
    rpc CreateRelationship(CreateRelationshipRequest) returns (CreateRelationshipResponse);

    // GetRelationship - Retrieve relationship by ID
    rpc GetRelationship(GetRelationshipRequest) returns (GetRelationshipResponse);

    // UpdateRelationship - Update relationship properties
    rpc UpdateRelationship(UpdateRelationshipRequest) returns (UpdateRelationshipResponse);

    // DeleteRelationship - Delete relationship
    rpc DeleteRelationship(DeleteRelationshipRequest) returns (DeleteRelationshipResponse);

    // FindRelationships - Find relationships by type and properties
    rpc FindRelationships(FindRelationshipsRequest) returns (FindRelationshipsResponse);

    // ========== Graph Traversal ==========

    // GetNeighbors - Get neighboring nodes
    rpc GetNeighbors(GetNeighborsRequest) returns (GetNeighborsResponse);

    // GetPath - Find path between two nodes
    rpc GetPath(GetPathRequest) returns (GetPathResponse);

    // GetShortestPath - Find shortest path between nodes
    rpc GetShortestPath(GetShortestPathRequest) returns (GetShortestPathResponse);

    // GetAllPaths - Find all paths between nodes
    rpc GetAllPaths(GetAllPathsRequest) returns (GetAllPathsResponse);

    // ========== Graph Algorithms ==========

    // PageRank - Run PageRank algorithm
    rpc PageRank(PageRankRequest) returns (PageRankResponse);

    // BetweennessCentrality - Calculate betweenness centrality
    rpc BetweennessCentrality(BetweennessCentralityRequest) returns (BetweennessCentralityResponse);

    // CommunityDetection - Detect communities in graph
    rpc CommunityDetection(CommunityDetectionRequest) returns (CommunityDetectionResponse);

    // ========== Schema Management ==========

    // CreateIndex - Create index on label/property
    rpc CreateIndex(CreateIndexRequest) returns (CreateIndexResponse);

    // DropIndex - Drop an index
    rpc DropIndex(DropIndexRequest) returns (DropIndexResponse);

    // ListIndexes - List all indexes
    rpc ListIndexes(ListIndexesRequest) returns (ListIndexesResponse);

    // CreateConstraint - Create uniqueness or existence constraint
    rpc CreateConstraint(CreateConstraintRequest) returns (CreateConstraintResponse);

    // DropConstraint - Drop a constraint
    rpc DropConstraint(DropConstraintRequest) returns (DropConstraintResponse);

    // ListConstraints - List all constraints
    rpc ListConstraints(ListConstraintsRequest) returns (ListConstraintsResponse);

    // ========== Transaction Support ==========

    // BeginTransaction - Start a new transaction
    rpc BeginTransaction(BeginTransactionRequest) returns (BeginTransactionResponse);

    // CommitTransaction - Commit transaction
    rpc CommitTransaction(CommitTransactionRequest) returns (CommitTransactionResponse);

    // RollbackTransaction - Rollback transaction
    rpc RollbackTransaction(RollbackTransactionRequest) returns (RollbackTransactionResponse);

    // ========== Database Operations ==========

    // GetDatabaseInfo - Get database information
    rpc GetDatabaseInfo(GetDatabaseInfoRequest) returns (GetDatabaseInfoResponse);

    // ListDatabases - List all databases
    rpc ListDatabases(ListDatabasesRequest) returns (ListDatabasesResponse);

    // GetStatistics - Get database statistics
    rpc GetStatistics(GetStatisticsRequest) returns (GetStatisticsResponse);

    // ========== Health Check ==========

    // HealthCheck - Service health check
    rpc HealthCheck(Neo4jHealthCheckRequest) returns (Neo4jHealthCheckResponse);
}

// ========================================
// Data Structures
// ========================================

// Node representation
message Node {
    int64 id = 1;                                    // Internal Neo4j node ID
    repeated string labels = 2;                      // Node labels
    google.protobuf.Struct properties = 3;           // Node properties
}

// Relationship representation
message Relationship {
    int64 id = 1;                                    // Internal Neo4j relationship ID
    int64 start_node_id = 2;                         // Start node ID
    int64 end_node_id = 3;                           // End node ID
    string type = 4;                                 // Relationship type
    google.protobuf.Struct properties = 5;           // Relationship properties
}

// Path representation
message Path {
    repeated Node nodes = 1;
    repeated Relationship relationships = 2;
    int32 length = 3;                                // Number of relationships in path
}

// Query result row
message ResultRow {
    map<string, google.protobuf.Value> fields = 1;   // Column name -> value mapping
}

// Query result summary
message QuerySummary {
    string query_type = 1;                           // READ_ONLY, READ_WRITE, WRITE_ONLY, SCHEMA_WRITE
    int32 nodes_created = 2;
    int32 nodes_deleted = 3;
    int32 relationships_created = 4;
    int32 relationships_deleted = 5;
    int32 properties_set = 6;
    int32 labels_added = 7;
    int32 labels_removed = 8;
    int32 indexes_added = 9;
    int32 indexes_removed = 10;
    int32 constraints_added = 11;
    int32 constraints_removed = 12;
    double execution_time_ms = 13;
}

// ========================================
// Cypher Query Operations
// ========================================

message RunCypherRequest {
    common.RequestMetadata metadata = 1;

    string cypher = 2;                               // Cypher query
    map<string, google.protobuf.Value> parameters = 3; // Query parameters
    optional string database = 4;                    // Target database (default: neo4j)
    optional int32 timeout_ms = 5;                   // Query timeout
}

message RunCypherResponse {
    common.ResponseMetadata metadata = 1;

    repeated ResultRow rows = 2;                     // Result rows
    repeated string columns = 3;                     // Column names
    QuerySummary summary = 4;                        // Query execution summary
}

message RunCypherReadRequest {
    common.RequestMetadata metadata = 1;

    string cypher = 2;
    map<string, google.protobuf.Value> parameters = 3;
    optional string database = 4;
    optional int32 timeout_ms = 5;
}

message RunCypherReadResponse {
    common.ResponseMetadata metadata = 1;

    repeated ResultRow rows = 2;
    repeated string columns = 3;
    QuerySummary summary = 4;
}

message RunCypherWriteRequest {
    common.RequestMetadata metadata = 1;

    string cypher = 2;
    map<string, google.protobuf.Value> parameters = 3;
    optional string database = 4;
    optional int32 timeout_ms = 5;
}

message RunCypherWriteResponse {
    common.ResponseMetadata metadata = 1;

    repeated ResultRow rows = 2;
    repeated string columns = 3;
    QuerySummary summary = 4;
}

message RunCypherBatchRequest {
    common.RequestMetadata metadata = 1;

    repeated CypherQuery queries = 2;
    optional string database = 3;
}

message CypherQuery {
    string cypher = 1;
    map<string, google.protobuf.Value> parameters = 2;
}

message RunCypherBatchResponse {
    common.ResponseMetadata metadata = 1;

    repeated RunCypherResponse results = 2;
}

// ========================================
// Node Operations
// ========================================

message CreateNodeRequest {
    common.RequestMetadata metadata = 1;

    repeated string labels = 2;                      // Node labels
    google.protobuf.Struct properties = 3;           // Node properties
    optional string database = 4;
}

message CreateNodeResponse {
    common.ResponseMetadata metadata = 1;

    Node node = 2;                                   // Created node
}

message GetNodeRequest {
    common.RequestMetadata metadata = 1;

    int64 node_id = 2;
    optional string database = 3;
}

message GetNodeResponse {
    common.ResponseMetadata metadata = 1;

    Node node = 2;
    bool found = 3;
}

message UpdateNodeRequest {
    common.RequestMetadata metadata = 1;

    int64 node_id = 2;
    google.protobuf.Struct properties = 3;           // Properties to set/update
    optional bool replace = 4;                       // Replace all properties (default: merge)
    optional string database = 5;
}

message UpdateNodeResponse {
    common.ResponseMetadata metadata = 1;

    Node node = 2;
}

message DeleteNodeRequest {
    common.RequestMetadata metadata = 1;

    int64 node_id = 2;
    optional bool detach = 3;                        // Delete relationships too
    optional string database = 4;
}

message DeleteNodeResponse {
    common.ResponseMetadata metadata = 1;

    bool success = 2;
    int32 relationships_deleted = 3;
}

message MergeNodeRequest {
    common.RequestMetadata metadata = 1;

    repeated string labels = 2;
    google.protobuf.Struct match_properties = 3;     // Properties to match on
    google.protobuf.Struct set_properties = 4;       // Properties to set if created/matched
    optional string database = 5;
}

message MergeNodeResponse {
    common.ResponseMetadata metadata = 1;

    Node node = 2;
    bool created = 3;                                // True if node was created, false if matched
}

message FindNodesRequest {
    common.RequestMetadata metadata = 1;

    optional string label = 2;                       // Filter by label
    google.protobuf.Struct properties = 3;           // Filter by properties
    optional int32 limit = 4;
    optional int32 skip = 5;
    optional string database = 6;
}

message FindNodesResponse {
    common.ResponseMetadata metadata = 1;

    repeated Node nodes = 2;
    int32 total_count = 3;
}

// ========================================
// Relationship Operations
// ========================================

message CreateRelationshipRequest {
    common.RequestMetadata metadata = 1;

    int64 start_node_id = 2;
    int64 end_node_id = 3;
    string type = 4;                                 // Relationship type
    google.protobuf.Struct properties = 5;
    optional string database = 6;
}

message CreateRelationshipResponse {
    common.ResponseMetadata metadata = 1;

    Relationship relationship = 2;
}

message GetRelationshipRequest {
    common.RequestMetadata metadata = 1;

    int64 relationship_id = 2;
    optional string database = 3;
}

message GetRelationshipResponse {
    common.ResponseMetadata metadata = 1;

    Relationship relationship = 2;
    bool found = 3;
}

message UpdateRelationshipRequest {
    common.RequestMetadata metadata = 1;

    int64 relationship_id = 2;
    google.protobuf.Struct properties = 3;
    optional bool replace = 4;
    optional string database = 5;
}

message UpdateRelationshipResponse {
    common.ResponseMetadata metadata = 1;

    Relationship relationship = 2;
}

message DeleteRelationshipRequest {
    common.RequestMetadata metadata = 1;

    int64 relationship_id = 2;
    optional string database = 3;
}

message DeleteRelationshipResponse {
    common.ResponseMetadata metadata = 1;

    bool success = 2;
}

message FindRelationshipsRequest {
    common.RequestMetadata metadata = 1;

    optional int64 start_node_id = 2;                // Filter by start node
    optional int64 end_node_id = 3;                  // Filter by end node
    optional string type = 4;                        // Filter by type
    google.protobuf.Struct properties = 5;
    optional int32 limit = 6;
    optional string database = 7;
}

message FindRelationshipsResponse {
    common.ResponseMetadata metadata = 1;

    repeated Relationship relationships = 2;
    int32 total_count = 3;
}

// ========================================
// Graph Traversal
// ========================================

message GetNeighborsRequest {
    common.RequestMetadata metadata = 1;

    int64 node_id = 2;
    optional string relationship_type = 3;           // Filter by relationship type
    optional string direction = 4;                   // INCOMING, OUTGOING, BOTH
    optional int32 max_depth = 5;                    // Traversal depth (default: 1)
    optional string database = 6;
}

message GetNeighborsResponse {
    common.ResponseMetadata metadata = 1;

    repeated Node nodes = 2;
    repeated Relationship relationships = 3;
}

message GetPathRequest {
    common.RequestMetadata metadata = 1;

    int64 start_node_id = 2;
    int64 end_node_id = 3;
    optional string relationship_type = 4;
    optional int32 max_depth = 5;
    optional string database = 6;
}

message GetPathResponse {
    common.ResponseMetadata metadata = 1;

    Path path = 2;
    bool found = 3;
}

message GetShortestPathRequest {
    common.RequestMetadata metadata = 1;

    int64 start_node_id = 2;
    int64 end_node_id = 3;
    optional string relationship_type = 4;
    optional int32 max_depth = 5;
    optional string database = 6;
}

message GetShortestPathResponse {
    common.ResponseMetadata metadata = 1;

    Path path = 2;
    bool found = 3;
}

message GetAllPathsRequest {
    common.RequestMetadata metadata = 1;

    int64 start_node_id = 2;
    int64 end_node_id = 3;
    optional string relationship_type = 4;
    optional int32 max_depth = 5;
    optional int32 limit = 6;
    optional string database = 7;
}

message GetAllPathsResponse {
    common.ResponseMetadata metadata = 1;

    repeated Path paths = 2;
}

// ========================================
// Graph Algorithms
// ========================================

message PageRankRequest {
    common.RequestMetadata metadata = 1;

    optional string label = 2;                       // Node label to run on
    optional string relationship_type = 3;
    optional int32 max_iterations = 4;
    optional double dampening_factor = 5;
    optional string database = 6;
}

message PageRankResponse {
    common.ResponseMetadata metadata = 1;

    map<int64, double> scores = 2;                   // Node ID -> PageRank score
}

message BetweennessCentralityRequest {
    common.RequestMetadata metadata = 1;

    optional string label = 2;
    optional string relationship_type = 3;
    optional string database = 4;
}

message BetweennessCentralityResponse {
    common.ResponseMetadata metadata = 1;

    map<int64, double> scores = 2;                   // Node ID -> centrality score
}

message CommunityDetectionRequest {
    common.RequestMetadata metadata = 1;

    optional string label = 2;
    optional string relationship_type = 3;
    optional string algorithm = 4;                   // louvain, label_propagation, weakly_connected
    optional string database = 5;
}

message CommunityDetectionResponse {
    common.ResponseMetadata metadata = 1;

    map<int64, int64> communities = 2;               // Node ID -> Community ID
    int32 num_communities = 3;
}

// ========================================
// Schema Management
// ========================================

message CreateIndexRequest {
    common.RequestMetadata metadata = 1;

    string index_name = 2;
    string label = 3;
    repeated string properties = 4;
    optional string index_type = 5;                  // BTREE, FULLTEXT, LOOKUP
    optional string database = 6;
}

message CreateIndexResponse {
    common.ResponseMetadata metadata = 1;

    bool success = 2;
}

message DropIndexRequest {
    common.RequestMetadata metadata = 1;

    string index_name = 2;
    optional string database = 3;
}

message DropIndexResponse {
    common.ResponseMetadata metadata = 1;

    bool success = 2;
}

message ListIndexesRequest {
    common.RequestMetadata metadata = 1;

    optional string database = 2;
}

message ListIndexesResponse {
    common.ResponseMetadata metadata = 1;

    repeated IndexInfo indexes = 2;
}

message IndexInfo {
    string name = 1;
    string type = 2;
    repeated string labels_or_types = 3;
    repeated string properties = 4;
    string state = 5;                                // ONLINE, POPULATING, FAILED
}

message CreateConstraintRequest {
    common.RequestMetadata metadata = 1;

    string constraint_name = 2;
    string label = 3;
    repeated string properties = 4;
    string constraint_type = 5;                      // UNIQUE, EXISTS, NODE_KEY
    optional string database = 6;
}

message CreateConstraintResponse {
    common.ResponseMetadata metadata = 1;

    bool success = 2;
}

message DropConstraintRequest {
    common.RequestMetadata metadata = 1;

    string constraint_name = 2;
    optional string database = 3;
}

message DropConstraintResponse {
    common.ResponseMetadata metadata = 1;

    bool success = 2;
}

message ListConstraintsRequest {
    common.RequestMetadata metadata = 1;

    optional string database = 2;
}

message ListConstraintsResponse {
    common.ResponseMetadata metadata = 1;

    repeated ConstraintInfo constraints = 2;
}

message ConstraintInfo {
    string name = 1;
    string type = 2;
    string label = 3;
    repeated string properties = 4;
}

// ========================================
// Transaction Support
// ========================================

message BeginTransactionRequest {
    common.RequestMetadata metadata = 1;

    optional string database = 2;
    optional int32 timeout_ms = 3;
}

message BeginTransactionResponse {
    common.ResponseMetadata metadata = 1;

    string transaction_id = 2;
}

message CommitTransactionRequest {
    common.RequestMetadata metadata = 1;

    string transaction_id = 2;
}

message CommitTransactionResponse {
    common.ResponseMetadata metadata = 1;

    bool success = 2;
}

message RollbackTransactionRequest {
    common.RequestMetadata metadata = 1;

    string transaction_id = 2;
}

message RollbackTransactionResponse {
    common.ResponseMetadata metadata = 1;

    bool success = 2;
}

// ========================================
// Database Operations
// ========================================

message GetDatabaseInfoRequest {
    common.RequestMetadata metadata = 1;

    optional string database = 2;
}

message GetDatabaseInfoResponse {
    common.ResponseMetadata metadata = 1;

    string name = 2;
    int64 node_count = 3;
    int64 relationship_count = 4;
    int64 label_count = 5;
    int64 relationship_type_count = 6;
    int64 property_key_count = 7;
}

message ListDatabasesRequest {
    common.RequestMetadata metadata = 1;
}

message ListDatabasesResponse {
    common.ResponseMetadata metadata = 1;

    repeated DatabaseInfo databases = 2;
}

message DatabaseInfo {
    string name = 1;
    string status = 2;                               // online, offline
    bool default = 3;
}

message GetStatisticsRequest {
    common.RequestMetadata metadata = 1;

    optional string database = 2;
}

message GetStatisticsResponse {
    common.ResponseMetadata metadata = 1;

    google.protobuf.Struct statistics = 2;
}

// ========================================
// Health Check
// ========================================

message Neo4jHealthCheckRequest {
    common.RequestMetadata metadata = 1;
}

message Neo4jHealthCheckResponse {
    common.ResponseMetadata metadata = 1;

    bool healthy = 2;
    string version = 3;
    string edition = 4;                              // community, enterprise
}
